torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -451,77 +456,212 @@ void FusionKernelRuntime::prepareRuntimeOrder() {
   }
 }
 
-std::vector<at::Tensor> FusionKernelRuntime::runWithInput(
-    const at::ArrayRef<IValue>& inputs,
-    size_t input_id) {
-  FUSER_PERF_SCOPE("FusionKernelRuntime::runMultiKernelWithInput");
+// passing args by value, since we will be modify this
+void FusionKernelRuntime::compileAsync(KernelArgumentHolder& args_old) {
+  // only single compilation is supported at this moment.
+  std::unique_lock<std::mutex> unique_lock(mutex_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported");
+  std::unique_lock<std::mutex> unique_lock2(compiling_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock2.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported 2");
+
+  // for some reason I can't seem to move unique_lock and it keeps using copy.
+  // auto func = [args = std::move(args_old), lock = std::move(unique_lock),
+  // this] () mutable {
+  auto func = [args = std::move(args_old), this]() mutable {
+    std::lock_guard<std::mutex> guard(compiling_);
+
+    // locking mutex_ since we are touching executors_ during compilation.
+    // c10::DeviceGuard dg(c10::Device(DeviceType::CUDA,
+    // args.getDeviceIndex())); CUDAGuard uses runtime API directly, which is
+    // thread safe.
+    c10::cuda::CUDAGuard dg(args.getDeviceIndex());
+
+    FUSER_PERF_SCOPE("FusionKernelRuntime::compileAsync");
 
-  TORCH_INTERNAL_ASSERT(
-      inputs.size() == segmented_fusion_->inputs().size(),
-      "Inputs were not set up correctly, recieved ",
-      inputs.size(),
-      " inputs but expecting ",
-      segmented_fusion_->inputs().size());
+    TORCH_INTERNAL_ASSERT(
+        args.size() == segmented_fusion_->inputs().size(),
+        "Inputs were not set up correctly, recieved ",
+        args.size(),
+        " inputs but expecting ",
+        segmented_fusion_->inputs().size());
+
+    c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
+    std::unordered_map<Val*, const ArgAbstract*> tensor_map;
+    mapArguments(tensor_map, args);
+
+    for (auto group_to_run : runtime_workspace_.group_run_order) {
```

Comment:
That's a very good point~ We likely can do it in parallel, since we only needed expression evaluation to get output sizes, which doesn't require a compiled kernel. I'm putting a TODO here as well.